File: tests/Graphics/Sprout.Graphics.Tests.SimpleShader/SimpleShaderTest.cs
using System.Drawing;
using Sprout.Graphics.ShaderUtils;
using Sprout.Graphics.Tests.Base;

namespace Sprout.Graphics.Tests.SimpleShader;

public class SimpleShaderTest() : TestBase("Simple Shader Test")
{
    private Shader _shader = null!;
    private Renderable _renderable = null!;
    
    protected override void Load()
    {
        string hlsl = File.ReadAllText("Shader.hlsl");
        byte[] vtx = Compiler.TranspileHLSL(Device.Backend, ShaderStage.Vertex, hlsl, "VSMain");
        byte[] pxl = Compiler.TranspileHLSL(Device.Backend, ShaderStage.Pixel, hlsl, "PSMain");

        ShaderInfo shaderInfo = new()
        {
            VertexShader = new ShaderAttachment(vtx, "VSMain"),
            PixelShader = new ShaderAttachment(pxl, "PSMain")
        };
        
        _shader = Device.CreateShader(in shaderInfo);

        RenderableInfo info = new()
        {
            Shader = _shader
        };
        _renderable = Device.CreateRenderable(in info);
    }

    protected override void Loop(float dt)
    {
        Device.Clear(Color.CornflowerBlue);
        _renderable.Draw(3);
        Device.Present();
    }

    public override void Dispose()
    {
        _renderable.Dispose();
        _shader.Dispose();
        
        base.Dispose();
    }
}

File: tests/Graphics/SimpleShaderTest.cpp
#include "Base/TestBase.h"

const char* Shader = R"(
float4 VSMain(const in uint id: SV_VertexID): SV_Position
{
    // Triangle
    const float2 vertices[] =
    {
        float2(-0.5, -0.5),
        float2( 0.0,  0.5),
        float2( 0.5, -0.5),
    };

    return float4(vertices[id], 0.0, 1.0);
}


float4 PSMain(): SV_Target0
{
    return float4(1.0, 0.5, 0.25, 1.0);
}
)";

class SimpleShaderTest : public TestBase
{
    std::unique_ptr<Sprout::Shader> _shader;

    void Load() override
    {

    }

    void Loop(float dt) override
    {
        auto& device = Device();
        device.Clear({ 0.25f, 0.5f, 1.0f });
        device.Present();
    }
};

int main(int argc, char* argv[])
{
    SimpleShaderTest test;
    test.Run("Simple Shader Test");

    return 0;
}